Create option flow for Rfxtrx integration

RobBie1221 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
- yaml support is limited to one-time import
- Options must be set through the UI once import is done (global options/add device/remove device/change device options)
- For newly added or detected devices, device_class cannot be set. Instead, it should be set by customizing entities:
https://www.home-assistant.io/docs/configuration/customizing-devices/

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Implement an option flow for Rfxtrx integration to:
- Change global options
- Change device options
- Add new devices by id

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14158

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RobBie1221 committed:
Initial setup for a possible options flow

RobBie1221 committed:
Reduce options and reorganize

RobBie1221 commented:
@elupus Here you go. I think this PR should address the following:

- Import yaml only first time, then ignore (to prevent changed options through flow to become inconsistent with yaml
- Change general options
- Change device options
- Add device by id

Next step would be (for a separate PR) to add a config flow to be able to add the integration without use of yaml.

About the device options, at least the device_class, if I understand correctly, it is only relevant for binary_sensor right? How about the other options? Do they make sense for other platforms than the binary_sensor?

RobBie1221 commented:
Perhaps we should re-think also code below. Are we not abusing the input scheme here? I just put in 2 options here but there are potentially more (e.g. handle case where a device starts transmitting with a new id). Generally my issue here is that ideally the strings are obtained through localization, but I have not (yet) found a way to do that. Aside from that, choosing what option function to launch is quite user friendly.

The alternative would be to have an empty "new id" field and selector for current configured devices and when the user either fills something in or selects an existing device, show a new window with options.

`options = ["Add device by id", "Configure device options"]`

`options_scheme = {
            vol.Optional(CONF_DEBUG): bool,
            vol.Optional(CONF_AUTOMATIC_ADD): bool,
            vol.Optional("selected_option"): vol.In(options),
        }`

elupus commented:
I think device class can be auto detected in most cases actually. And it can be set from customize.yamlm 

I will look it over later.

elupus commented:
flow:
* I don't think we should mix settings and the choice on what to do. So start of with an input selector with global options, device config, add device. I don't think we can localize the values yet.

* adding devices likely need to be by event string (your is int)

device_class:
I started on auto detection of device class here: https://github.com/home-assistant/core/pull/38000. But in general i think we should drop the ability to set this here. You can set it from customize.yaml.

RobBie1221 commented:
That was exactly why I second guessed, mixing settings and what to do. Perhaps in time an option flow will get options to trigger specific option functions, but for now I'll get rid of it.

You're right about adding devices, this won't work with int.

I'll get rid of the device_class, saves some trouble as well to retrieve all possible classes where it's really not needed when you can auto-detect and users can always use customization to override.

I'll have a look at your PR this evening.

RobBie1221 committed:
Rework options in options flow

RobBie1221 committed:
Work in progress

RobBie1221 commented:
@elupus Now that the ConfigEntry contains all info I thought I had everything in place, but I'm getting stuck on the next item. The ConfigEntry data is still in the same scheme as the yaml is. That means devices are stored by event_code.

The issue I'm having, the options flow can quite easily get a list of devices, but I wouldn't know easily how to link it to the key in the ConfigEntry. As far as I can see, this exact key is nowhere stored? For configured devices I could off course use functions to come to a device id but this is cumbersome and expensive. For detected devices I'm not sure if the key is stored at all.

I was thinking in the line off changing the ConfigEntry to save both device id and event_code together. It seems that the event code is used upon initialization so not sure if it's a good idea to remove it completely?
That leaves the issue for detected devices. Perhaps these can be stored in hass.data for new devices (store event_code+device id) and once a user configures options through option flow add it to the ConfigEntry?

What are your thoughts about this? I'm a bit hesitating to put in a lot of lines of code and then have a discussion about the best way to do it :)

elupus commented:
Yup. herein lies dragons :). The problem is the event really is defining the device. So it probably need to be there. The device_id should likely be there too. That said, should we perhaps move the whole device config to the optionflow storage from the start (however you do that i don't know).

And i think whenever we detect a new device, we should trigger an automatic update of the option entry. Ie an option flow that never shows any dialog.

RobBie1221 commented:
Perhaps it's sufficient to restrict to device_id and event_code. Storing the whole object would make it difficult to have changes there. These two items would imho be sufficient as we can link devices to event_codes and vice versa. Probably shouldn't make ConfigEntry items too big as well.

I was thinking for the options flow to just trigger a reload_entry (which is why I needed the unload). This makes sense for almost everything that is done in the options flow except what you mention, just triggering a flow to add a key for a device that has already been added but needs a registration in the entry. I'll try to find out what our options here are.

I think I will make a separate PR to change the ConfigEntry a bit.

RobBie1221 commented:
Perhaps one more, I've raised the point of sensors which start reporting with another ID before. I'm not even nearly there to start working on that, but perhaps globally we can think of how this could look like and see if it affects things we are changing here.

Basically I have below example. These are 4+3 temp/hum sensors, same brand 2 different models. I can set a channel on the device which probably changes part of the code, another part is randomly generated when putting the batteries in.

Do you have in mind how a migrate function could work? Is there any way you see could work where we keep the existing device+entities? Obviously the issue is they link to the "old" device with probably keys like unique_id and device_id. Is there any way to delete a device+entities but copy their user configured properties to new ones?
Another way could be to link the new code to existing device and entities, but that's a bit praying a new device won't get an id that another device used to have (which I think doesn't actually happen but there are no guarantees here).

```
rfxtrx:
  device: /dev/ttyUSB1
  automatic_add: True
  devices:
    0a520101f00400e22d0189:
    0a520110ff0400df400169:
    0a520101f50400cc260279:
    0a520105230400c3260279:
    0a520202fd0700c32b0069:
    0a520204bd0200d6270279:    
    0a520207e80300d6270269:
```

elupus commented:
i think devices are allowed to have multiple identifiers. so we could add additional. then let entities grab its unique id from device registry.

but that would over time be problematic with reuse of ids. so update and replace is probably needed.

RobBie1221 commented:
Given the fact that devices can have multiple identifiers, it would also be possible to store the event_code in the device id.

This would be an easy way to link devices to ConfigEntry but I think it would break things for users on existing devices right? The upside would be that it's just one extra entry. The downside of the alternative (store device id in ConfigEntry) is that it adds quite some complexity, inside ConfigFlow and OptionFlow it should convert to device id and store and discovered devices must also somehow trigger a flow to add ConfigEntry items. The plus side of this is that at least for every device there is already a ConfigEntry, although I don't know if this really brings something.

RobBie1221 commented:
Never mind my comment above, no need to trigger the OptionsFlow to add a newly detected device, we can call async_update_entry to add these to the ConfigEntry. Adding to ConfigEntry then makes sense, should be easy and the entry is always is sync with devices in the device registry.

RobBie1221 commented:
@elupus You plan to drop CONF_DEBUG right? I saw your PR in the pyRFXtrx repo. I'm refactoring the ConfigEntry to have data that need to be set in the options flow move to the options key in the ConfigEntry, but the CONF_DEBUG will give me some additional refactoring which is not really needed when we want to drop this (we can leave it in data for now)

elupus commented:
Yes I think it's rather pointless. So you can ignore that key from option flow.

RobBie1221 commented:
Hmm...it was a good idea (and still is) to move things from data to options in the ConfigEntry, but it's quite impossible to get things in the options key during import. That makes this exercise rather pointless. I'll leave the CONF_DEBUG in, once the options flow is in it can't be set through yaml either after first import. We just have to remember when bumping to a new pyRFXtrx to remove this from options flow.

RobBie1221 committed:
Add logic to update config and reload integration

RobBie1221 committed:
Implement correct device options

RobBie1221 committed:
Implement logic for setting options

RobBie1221 committed:
Adjust test import update

RobBie1221 committed:
Make pylint happy

RobBie1221 committed:
Finish basic option flow

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,251 @@
+    """Handle Rfxtrx options."""
+
+    def __init__(self, config_entry: ConfigEntry) -> None:
+        """Initialize vizio options flow."""
```
Stale comment

RobBie1221 replied:
Good old copy/paste error...I actually pasted it from your rfxtrx_option_flow branch 😆 

I'll remove it :)

elupus reviewed (commented):
Not on computer, will look more later.

RobBie1221 commented:
Thanks. the basic logic is done, I'm working on 2 minor additions which I will push in a moment:
1. Remove device option: now that devices always end up in a ConfigEntry it's impossible to remove a device, so we'll need an option to remove it from the ConfigEntry
2. Fix strings.json and en.json

There is one more thing that is not so nice, but perhaps it would be good to make a separate PR for that not to make too much of a spaghetti of this one. When PT2262 devices have been added from discovery probably multiple devices were created. When one device is selected and configuration for `data_bits` is added, the `device_id` changed so we need to migrate the existing device and can clean up additional devices created.

RobBie1221 commented:
And test code is not yet done...probably CI will fail on codecov

I also added myself as codeowner if you don't mind.

elupus commented:
Right yea test is needed. Yep make perfect sense to get you in ad code owner (the I can get some backup in the flack ive been getting in issues ;)

elupus commented:
I regret not leaving naming in until later. Would have been easy to keep optional, and skip for option flow. Would have simplified many people migrations.

RobBie1221 commented:
Yeah regrets. Don't waste too much time on them. I think renaming becomes easier once option flow is in, now people follow a strategy of adding one device, restarting ha, renaming, adding another etc. With option flow they can add through option flow, rename etc. No need to restart every time.

elupus commented:
We need to fix init i just realised by some other comment. We register listeners based on start event, that won't work in restart of entry.
We can't just pass it to constructor of Connect() unless we guard it in the callback with a asyncio.Lock(). Since events comes in before the constructor returns.

But i think we can guard in the callback, but still set it the way we do after.

RobBie1221 commented:
Ah crap, missed that. Would you want to refactor that here or make a separate PR?

elupus commented:
We do a separate pull

RobBie1221 committed:
Get started with remove device

RobBie1221 committed:
Fix small issue

RobBie1221 commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,273 @@
+                    errors = {"base": "unknown_event_code"}
+                else:
+                    device_registry = await async_get_registry(self.hass)
+                    device_registry.async_remove_device(user_input[CONF_REMOVE_DEVICE])
```
This needs some work. This removes the device and entities, but the platform doesn't know it's been removed so it will try to unload entities when async_entry_unload is called.

MartinHjelmare replied:
Maybe move the clean up to the dispatch target instead?

RobBie1221 replied:
The dispatch target(s) are the entities and directly calls `self.async_remove` on the entities.

To have it in a dispatch target I'd have to add another listener for that during setup of the entry (and remove it during unload), which is why I did it here. Do you see that as a better solution?

RobBie1221 commented:
codecov passes, I noticed that the whole rfxtrx folder is added to .coveragerc. I know you did a lot on test code in the past weeks, is that entry still needed?

MartinHjelmare commented:
At least the config flow module should not be excluded from coverage calculation.

RobBie1221 commented:
The coverage is not that bad (90%-96%), except for config_flow, because I didn't make tests yet. I'll change .coveragerc here to exclude everything except config_flow and once this PR is done check if we can add some tests to get coverage >95.7% and remove it entirely from exclude list.

RobBie1221 committed:
Add platforms to .coveragerc

RobBie1221 committed:
Call unload on entities after removing device

RobBie1221 committed:
Tweak strings

RobBie1221 committed:
Remove commented code

RobBie1221 committed:
Add some tests

RobBie1221 committed:
Add more tests

RobBie1221 commented:
I removed the Draft, this is pretty much done from my side and ready for review. I need to work on a docs PR, will link it here once that is done.

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+                    event_code = device_data["event_code"]
+                    device_id = device_data["device_id"]
+                    self.hass.helpers.dispatcher.async_dispatcher_send(
+                        f"{DOMAIN}_{device_id}"
```
this is a very non descriptive event. Need a const value that describe what it is.

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+                self._rfxobj.device, data_bits=user_input.get(CONF_DATA_BITS)
+            )
+            try:
+                command_on = (
+                    int(user_input.get(CONF_COMMAND_ON), 16)
+                    if user_input.get(CONF_COMMAND_ON)
+                    else None
+                )
+                command_off = (
+                    int(user_input.get(CONF_COMMAND_OFF), 16)
+                    if user_input.get(CONF_COMMAND_OFF)
+                    else None
+                )
```
I'm not too fond of those inline if clauses. Something like this seems nicer. Event nicer to use voluptuous to validate.
 
```suggestion
                def none_or_int16(value):
                  if value is None:
                    return None
                  return int(value, 16)
 
                command_on = none_or_int16(user_input.get(CONF_COMMAND_ON))
                command_off = none_or_int16(user_input.get(CONF_COMMAND_OFF))
```

RobBie1221 replied:
I tried voluptuous, no success. Am I doing something wrong?

```
vol.Optional(
   CONF_COMMAND_ON,
   default=hex(device_data.get(CONF_COMMAND_ON, 0)),
   ): vol.All(str, lambda value: int(value, 16))
```

MartinHjelmare replied:
The voluptuous-serialize library we use to convert voluptuous for the frontend doesn't support complex validators.

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+
+        if user_input is not None:
+            device_id = get_device_id(
+                self._rfxobj.device, data_bits=user_input.get(CONF_DATA_BITS)
```
maybe _rfxobj should be called _selected_device_object?

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+
+            if not errors:
+                devices = {}
+                devices[self._selected_device_event_code] = {
+                    CONF_DEVICE_ID: device_id,
+                    CONF_FIRE_EVENT: user_input.get(CONF_FIRE_EVENT, False),
+                    CONF_SIGNAL_REPETITIONS: user_input.get(CONF_SIGNAL_REPETITIONS, 1),
+                }
+
+                device = devices[self._selected_device_event_code]
```
```suggestion
               device = {
                    CONF_DEVICE_ID: device_id,
                    CONF_FIRE_EVENT: user_input.get(CONF_FIRE_EVENT, False),
                    CONF_SIGNAL_REPETITIONS: user_input.get(CONF_SIGNAL_REPETITIONS, 1),
                }

                devices[self._selected_device_event_code] = device
```

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+                if (
+                    user_input.get(CONF_OFF_DELAY_ENABLED)
+                    and user_input.get(CONF_OFF_DELAY) is not None
+                ):
```
Why do we need delay_enabled?

RobBie1221 replied:
If off_delay must be disabled, it must be None. A value of "0" for example still means it is enabled and used. It seemed cleaner to give the user the option to use it and define value in separate input items.

Alternatively we could define behavior as disable means remove the input from the field, but I have to check if that can even work, not sure about that.

RobBie1221 replied:
@elupus I'm not getting it to work without. If I remove the default "0"
`vol.Optional(CONF_OFF_DELAY, default=device_data.get(CONF_OFF_DELAY)): int,`
And have a device which has off_delay, if I remove the input from the form it says "user input malformed" (even though it's `vol.Optional`. If I leave in the 0 and then remove the value it accepts, but the returned data contains the key with value '0'. I have no options to distinguish between someone wanting `off_delay` of `0` and someone wanting to disable `off_delay`

I think we need this delay_enabled.

MartinHjelmare replied:
Can we use suggested value instead of default?

https://developers.home-assistant.io/docs/data_entry_flow_index#show-form

RobBie1221 replied:
I tried it before, it's not working because when `off_delay` is configured and when you set the suggested value and then clear the field in the form, the `user_input` contains the field with `0` value. That wouldn't give me a way to distinguish between `0` and disabled.

It did got me thinking about an alternative strategy, what seems to work is not to define the field as `int`, but as `str`. When using the suggested value and clearing the field, the `user_input` does not contain the field. The downside is that users can enter any character in the field and will only get a parsing error after submitting, but it could work. It would need some logic, because e.g. code below gives a string "None" in the field, so we only should set suggested value when there is an off_delay.

```
vol.Optional(
  CONF_OFF_DELAY,
  description={
    "suggested_value": str(device_data.get(CONF_OFF_DELAY))
  },
```

RobBie1221 replied:
Resolving this, removed `off_delay_enabled` and went for the `str` solution. Has a small setback that users can enter wrong data, but they get a clear message.

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+                    and user_input.get(CONF_OFF_DELAY) is not None
+                ):
+                    device[CONF_OFF_DELAY] = user_input[CONF_OFF_DELAY]
+                if user_input.get(CONF_DATA_BITS) is not None:
```
What about empty values here? If you start typing something and erase it?

```suggestion
                if user_input.get(CONF_DATA_BITS):
```

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+                    device[CONF_OFF_DELAY] = user_input[CONF_OFF_DELAY]
+                if user_input.get(CONF_DATA_BITS) is not None:
+                    device[CONF_DATA_BITS] = user_input.get(CONF_DATA_BITS)
+                if command_on is not None:
```
```suggestion
                if command_on:
```

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+                    device[CONF_DATA_BITS] = user_input.get(CONF_DATA_BITS)
+                if command_on is not None:
+                    device[CONF_COMMAND_ON] = command_on
+                if command_off is not None:
```
```suggestion
                if command_off:
```

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+        device_data = self._selected_device
+
+        if self._rfxobj is None:
+            self._rfxobj = get_rfx_object(self._selected_device_event_code)
```
Are we sure it's the same event code here? Can't we return to this option screen again? It's not that heavy to calculate so maybe just always calculate.

elupus replied:
Also see other comment. Might be better to make sure this object is set before we enter dialog instead of setting inside the dialog.

elupus commented on homeassistant/components/rfxtrx/translations/en.json:
```diff
@@ -0,0 +1,34 @@
+                "data": {
+                  "debug": "Enable debugging",
+                  "automatic_add": "Enable automatic add",
+                  "device_id": "Enter device id to add",
```
This is an event code right? So needs re-woring. Also the key should probably reflect that so we don't mix it up with our device_id tuple.

elupus commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,281 @@
+            if CONF_DEVICE_ID in user_input:
+                self._selected_device_event_code = user_input[CONF_DEVICE_ID]
+                self._selected_device = {}
+                return await self.async_step_set_device_options()
```
We should validate that the event_code is parsable here. So i think we should set the _rfxobj before we go to device_options.

RobBie1221 commented:
Thanks for the comments! I'll rework it.

RobBie1221 committed:
Rework review comments 1

RobBie1221 committed:
Rework review comments 2

RobBie1221 committed:
Rework review comments 3

RobBie1221 committed:
Fix mixup between device id and event code

RobBie1221 committed:
Fix strings

RobBie1221 committed:
fix tests, rewind conversion in voluptuous

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+                else:
+                    entry_data[CONF_DEVICES][event_code] = options
+        self.hass.config_entries.async_update_entry(self._config_entry, data=entry_data)
+        self.hass.async_create_task(async_update_options(self.hass, self._config_entry))
```
Register a listener instead.

https://developers.home-assistant.io/docs/config_entries_options_flow_handler#signal-updates

RobBie1221 replied:
I did not do that on purpose. The `async_update_entry` is also called when `automatic_add` is `True` and new devices are discovered. In that case it shouldn't reload the integration.

I did not find a way to distinguish between whether the update comes from the option flow or the event from the device.

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+                    self.hass.helpers.dispatcher.async_dispatcher_send(
+                        f"{DOMAIN}_{CONF_REMOVE_DEVICE}_{device_id}"
+                    )
+                    device_registry = await async_get_registry(self.hass)
```
We already get the device registry in the first pass of this method. We should be able to reuse that variable by saving it as an instance attribute.

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+            if CONF_DEVICE in user_input:
+                device_data = self._get_device_data(user_input[CONF_DEVICE])
+                if not device_data.get(CONF_EVENT_CODE):
+                    errors = {"base": "unknown_event_code"}
```
When can this happen? How can the user amend it?

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+
+        device_data = self._selected_device
+
+        data_scheme = {
```
`data_schema`

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+                ):
+                    device[CONF_OFF_DELAY] = user_input[CONF_OFF_DELAY]
+                if user_input.get(CONF_DATA_BITS):
+                    device[CONF_DATA_BITS] = user_input.get(CONF_DATA_BITS)
```
Don't use `dict.get` if we know the key is in the dict. Use `dict[key]`.

MartinHjelmare commented on tests/components/rfxtrx/test_config_flow.py:
```diff
@@ -44,4 +48,297 @@ async def test_import_update(hass):
+    assert entry.data["devices"]["0b1100cd0213c7f230010f71"]["signal_repetitions"] == 5
+    assert "delay_off" not in entry.data["devices"]["0b1100cd0213c7f230010f71"]
+
+    state = hass.states.get("binary_sensor.ac_213c7f2_48")
```
Where is the integration set up?

RobBie1221 replied:
The option flow sets it up because it triggers a reload

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+
+        data_scheme = {
+            vol.Optional(
+                CONF_FIRE_EVENT, default=device_data.get(CONF_FIRE_EVENT, False)
```
Why do we need an option for fire event? We could always fire an event.

RobBie1221 replied:
@elupus What do you think? Does it hurt to fire an event for all devices that are in the device list?

elupus replied:
From my perspective no, but balloob didn't want it. I think we should, it's not many events and it's much easier for user.

RobBie1221 replied:
@MartinHjelmare See #37565

I think in time, this event can go, as binary sensor and sensor have force_update now so they always fire state events when receiving data. We have however seen that some commands do not end up in entities, like Group On/Off. Before we remove this, we need to be sure that events are always mapped to an entity.

Perhaps not touch it now and see if we can improve event to entity mapping such that it can be removed altogether at a later point in time?

elupus replied:
Then we can question the need for binary sensor all together. But it is starting to get tricky for user to configure this. They need to know quite intricate details of rfxtrx.

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+                        default=device_data.get(CONF_OFF_DELAY) is not None,
+                    ): bool,
+                    vol.Optional(
+                        CONF_OFF_DELAY, default=device_data.get(CONF_OFF_DELAY, 0)
```
In the long term, I'd like this feature and option to be removed.

RobBie1221 replied:
I saw the discussions in earlier PR's. Will think about it, perhaps @elupus already has some ideas.

RobBie1221 replied:
@MartinHjelmare @elupus What if we remove it and refer in the docs to `template`? I think this feature is only really needed when a user wants to have meaningful value in the UI (other things can be done through state events or checking last_updated). The only real downside is that sensor.time has 1 minute accuracy.

```
binary_sensor:
  - platform: template
    sensors:
      bedroom_occupancy:
        friendly_name: Test Off Delay
        value_template: >-
          {% if (as_timestamp(now()) - as_timestamp(states.binary_sensor.ac_213c7f2_48.last_updated) < 10) %}
            {{ is_state('binary_sensor.ac_213c7f2_48', 'On') }}
          {% else %}
            {{ "false" }}
          {% endif %}
        entity_id: sensor.time

sensor:
  - platform: time_date
    display_options:
      - "time"
```

elupus replied:
The template sensor support delay_off itself....

RobBie1221 replied:
The way I read the docs, it does something different. I read it as a delay after an off state has been set that the state really goes to off, but it needs someone to tell off before really going off.

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,294 @@
+                        global_options=self._global_options, devices=devices
+                    )
+
+                    return self.async_create_entry(title="", data={})
```
I'm thinking maybe we should abort instead to avoid updating the entry twice.

RobBie1221 replied:
Hmm...can you verify? I think I'm hitting a core issue.

In `config_entries.py` in `async_finish_flow` it's doing:

```
if result["data"] is not None:
  self.hass.config_entries.async_update_entry(entry, options=result["data"])
```

However `_create_abort_data` in `data_entry_flow.py` does not write "data" so this throws a KeyError

```
@callback
def _create_abort_data(
    flow_id: str,
    handler: str,
    reason: str,
    description_placeholders: Optional[Dict] = None,
) -> Dict[str, Any]:
    """Return the definition of an external step for the user to take."""
    return {
        "type": RESULT_TYPE_ABORT,
        "flow_id": flow_id,
        "handler": handler,
        "reason": reason,
        "description_placeholders": description_placeholders,
    }
```

RobBie1221 replied:
Aside from this, we cannot use abort because it is visible in the UI.

![image](https://user-images.githubusercontent.com/31007358/89037461-cda81800-d33e-11ea-9e08-ada2cfa16cb3.png)

MartinHjelmare replied:
We can set the abort reason string to show.

@balloob should we fix the options flow manager async_finish_flow to handle abort, or should we do something else?

Normally we save data in the options flow, but in this case the idea is to update the config entry data in the options flow instead. That's unconventional and why we're hitting this issue.

RobBie1221 replied:
We can set a string, but it would still show "Aborted" instead of "Finished" right? It really needs to show the latter.

The problem is not really updating instead of saving. We have two issues:
1. We must reload integration after options flow has finished, which internally calls `async_update_entry`. The listener that would normally be used for that however cannot be used because we also trigger `async_update_entry` to keep administration of new devices that are detected from rfx events.
2. Through option flow the `options` key is set and not the `data` key. As we still support import, it's not that straightforward to get device data in the `options` key as it's impossible to set this key from either import or config flow.

I had thought about finding a solution for the second, but it still would not give me a full solution so it seemed not worth the effort at this time.

elupus replied:
i really think we should find a way to get data into the options fields instead. Can't we load it into .data on import. On first time we start an option flow, we copy all data over to the options field instead, once there is data in the options field, we use that as a source for data?

RobBie1221 replied:
We can do that, but what does it bring to do that?

We still have to call update entry and reload manually, because we can't use the listener.

To me it seems like adding a lot of code which only adds complexity.

More generally speaking, I've seen PR's where others get creative as well to do these kind of copy actions, shouldn't it be considered then to be able to set options from import and config flow?

MartinHjelmare replied:
We would not need to update the entry, since that's done when the options flow is finished. We would just need to reload the entry.

I think it's more in line with our current design to use options.data as a source of truth for the options. That's what we normally do right now.

Here's the related architecture issue. We shouldn't wait for that.
https://github.com/home-assistant/architecture/issues/377

RobBie1221 replied:
Ok, I thought we needed to update the entry before we call `self.hass.async_create_task(async_update_options(self.hass, self._config_entry))` but if we don't need to and can let the options flow update the entry it's better to do so.

We would need to wait for #38463. Without it there is no clean way to get the data in the options key in the first place.

Do note that the 0.113 release doesn't store options in the `data` key, but the current dev branch does. If we want to avoid making migration functions for ConfigEntries we may need some traction on that topic.

RobBie1221 replied:
If #38463 gets in, it seems better to make a separate PR to move data to `options` to get the ConfigEntry sorted.

RobBie1221 committed:
Rework review comments

RobBie1221 committed:
Rework off delay

RobBie1221 committed:
Extend option flow tests

RobBie1221 committed:
Remove event code check

RobBie1221 committed:
Increase test coverage

elupus commented:
Well my idea was that once we remove yaml and import. Our setup is more in
line with what is expected and can be simplified without changing how data
is stored.

That said, maybe its a very small change to allow writing options during
entry creation. Will do a quick check.



Den lör 1 aug. 2020 19:21Rob Bierbooms <notifications@github.com> skrev:

> *@RobBie1221* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/rfxtrx/config_flow.py
> <https://github.com/home-assistant/core/pull/37982#discussion_r463982042>:
>
> > +                if not device_data.get(CONF_EVENT_CODE):
> +                    errors = {"base": "unknown_event_code"}
> +                else:
> +                    event_code = device_data[CONF_EVENT_CODE]
> +                    device_id = device_data[CONF_DEVICE_ID]
> +                    self.hass.helpers.dispatcher.async_dispatcher_send(
> +                        f"{DOMAIN}_{CONF_REMOVE_DEVICE}_{device_id}"
> +                    )
> +                    device_registry = await async_get_registry(self.hass)
> +                    device_registry.async_remove_device(user_input[CONF_REMOVE_DEVICE])
> +                    devices = {event_code: None}
> +                    self.update_config_data(
> +                        global_options=self._global_options, devices=devices
> +                    )
> +
> +                    return self.async_create_entry(title="", data={})
>
> We can do that, but what does it bring to do that?
>
> We still have to call update entry and reload manually, because we can't
> use the listener.
>
> To me it seems like adding a lot of code which only adds complexity.
>
> More generally speaking, I've seen PR's where others get creative as well
> to do these kind of copy actions, shouldn't it be considered then to be
> able to set options from import and config flow?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/37982#discussion_r463982042>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADV5IAJSXQHIFDSLA4KGP3R6RFI3ANCNFSM4PBPF6DA>
> .
>

RobBie1221 committed:
Resolve mere conflicts

RobBie1221 commented:
If we want to use the option flow as it was designed, I would really like to focus first on how we can get rid of the manual entry update and reload. Once we move data to options, we actually update options twice, now we update first data then options.

elupus commented:
Indeed. How annoying.

On Tue, Aug 4, 2020 at 1:52 PM Rob Bierbooms <notifications@github.com>
wrote:

> *@RobBie1221* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/rfxtrx/config_flow.py
> <https://github.com/home-assistant/core/pull/37982#discussion_r464994144>:
>
> > +
> +        data_scheme = {
> +            vol.Optional(
> +                CONF_FIRE_EVENT, default=device_data.get(CONF_FIRE_EVENT, False)
> +            ): bool,
> +        }
> +
> +        if binary_supported(self._selected_device_object):
> +            data_scheme.update(
> +                {
> +                    vol.Optional(
> +                        CONF_OFF_DELAY_ENABLED,
> +                        default=device_data.get(CONF_OFF_DELAY) is not None,
> +                    ): bool,
> +                    vol.Optional(
> +                        CONF_OFF_DELAY, default=device_data.get(CONF_OFF_DELAY, 0)
>
> The way I read the docs, it does something different. I read it as a delay
> after an off state has been set that the state really goes to off, but it
> needs someone to tell off before really going off.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/37982#discussion_r464994144>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADV5IB563FVYUKGVMF6KLDR67ZAZANCNFSM4PBPF6DA>
> .
>

RobBie1221 commented:
@elupus I have a branch ready to move everything from `data` to `options`, once your PR passes I'll make a PR for that and I'll adjust this one. I'll leave the `force_update` and `off_delay` untouched for now, that should be separate PR anyway if we want to deprecate those.

Can you also add the label breaking-change here? I've updated the header of the PR.

MartinHjelmare commented:
Please clarify what we mean with customization in the breaking change paragraph. Also consider adding a sentence on how users can change options instead of changing configuration.yaml, ie via the options flow that we add here. It's important that we not only mention what is gone in the breaking change but how the user can cope with the breakage.

RobBie1221 commented:
I added some text. I put a link to the customizing entities page, it can be done through the UI so it's best for users to follow that link. I also added text about options (indeed, focus was too much on what isn't possible anymore).

I was thinking, while writing the docs, for new users merging in this PR will make it very confusing to set it up. They need to add something to `configuration.yaml` only to set it up (without any validation whether the device is actually there). It seems strange to have it in that way. We should really have a config flow as well. Would it be an idea to add a very basic config flow to this PR only to set the host/port and add be able to add it through UI? This will only be a few lines of code, nothing complex.

We can address the communication validation in a separate PR, which will be more work.

elupus commented:
No I think add basic user config make sense in this pull.

RobBie1221 commented:
What do you think...should we prompt a window with question whether to configure serial or host/port and then give the respective options or make one window with device/host/port and document to only use either device or host/port?

elupus commented:
I think a prompt would be clearer. That way it would be easier to handle enumeration of available serial devices should we want to.

MartinHjelmare commented:
Can we make the config flow a separate PR instead? This won't go out in 0.114 anyway. And if we need more time than 3 weeks, we should use a separate target branch that we can merge PRs into.

It's better to split work in as many small pieces as possible.

RobBie1221 commented:
Ok, I'll make a separate PR for that. I don't think we need a lot of time for a config flow, this essentially took a lot of effort because we had to get some things sorted to fit on the options flow. The config flow should be pretty straightforward.

RobBie1221 commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -16,8 +300,12 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
-        if entry and import_config.items() != entry.data.items():
-            self.hass.config_entries.async_update_entry(entry, data=import_config)
-            return self.async_abort(reason="already_configured")
+        await self.async_set_unique_id(DOMAIN)
```
I just realized we probably need migration anyway and up the version of the ConfigEntry.

Users with 0.113 have no data at all stored in the ConfigEntry, and if they want to upgrade to a release where this merges in, the ConfigEntry will not be updated from yaml. That's quite user unfriendly.

elupus replied:
Only if we stop importing. Right now we re-write then config entry stored based on yaml. As soon as we turn it into a one time import yes.

We could during the first version that imports once, check if the stored entry is an old version and just ignore it completely then. Since all data should have come from yaml in that case and we can just do a full import.

elupus commented:
Since my pull request for setting options from config flow seem to have derailed into meeting config/options I think we should forget about event using the options field of the entry for now.

RobBie1221 commented:
To add to that, #36537 takes the exact same approach as we do here (update data and return empty options) which is mentioned here:
https://github.com/home-assistant/architecture/issues/377
as a fine solution for now.

I think the options/data needs more discussion.

RobBie1221 commented:
@MartinHjelmare What do you think? Is this acceptable as is? The only open point is the data vs options. The approach taken here is similar as in #36537. Setting options from import has been dropped so that's not a way to go at this time.

MartinHjelmare commented:
Ok.

I think we should merge this into another target branch than dev branch since we want the config flow to go into the same release. That's safer and we then don't need to rush anything.

Please open a new PR based on dev branch that we'll use as target for this PR. Then rebase here on top of that branch and swap target branch here in the PR.

RobBie1221 commented:
Ok, makes sense. I'm almost following how to do this, should this new branch I need to target here exist in my forked repo or the home-assistant/core repo? I think the latter right?

MartinHjelmare commented:
Yeah, it might be easier to find the branch if it's in this repo. I can push a branch and you can open a PR from that branch, so you'll get the contributor credit when we merge.

MartinHjelmare commented:
Here's the branch, a clean copy of dev:
https://github.com/home-assistant/core/tree/rfxtrx-user-options-flow

RobBie1221 commented:
Ok this should be done then, changed target and resolved merge conflicts. I can open a draft PR for the new branch into dev once this is merged. I will also relink the docs PR to the new PR, that way the docs can be updated in one go as well.

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,296 @@
+                self._selected_device = self._config_entry.data[CONF_DEVICES][
+                    event_code
+                ]
+                self._selected_device_object = get_rfx_object(
+                    self._selected_device_event_code
+                )
```
```suggestion
                self._selected_device_object = get_rfx_object(event_code)
```

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,296 @@
+                )
+                return await self.async_step_set_device_options()
+            if CONF_REMOVE_DEVICE in user_input:
+                device_data = self._get_device_data(user_input[CONF_REMOVE_DEVICE])
```
```suggestion
                entry_id = user_input[CONF_REMOVE_DEVICE]
                device_data = self._get_device_data(entry_id)
```

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,296 @@
+                self.hass.helpers.dispatcher.async_dispatcher_send(
+                    f"{DOMAIN}_{CONF_REMOVE_DEVICE}_{device_id}"
+                )
+                self._device_registry.async_remove_device(
+                    user_input[CONF_REMOVE_DEVICE]
+                )
```
```suggestion
                self._device_registry.async_remove_device(entry_id)
```

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,296 @@
+                data_bits=user_input.get(CONF_DATA_BITS),
+            )
+
+            def none_or_int(value, base):
```
We can move this function to the module level.

RobBie1221 replied:
Where do you want to move it? (What would be the module level)

MartinHjelmare replied:
Either after the constants or at the end of this module.

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,296 @@
+        """Get event code based on device identifier."""
+        event_code = None
+        device_id = None
+        for entry in self._device_entries:
+            if entry.id == entry_id:
+                device_id = next(iter(entry.identifiers))[1:]
+                break
```
```suggestion
        entry = self._device_registry.async_get(entry_id)
        device_id = next(iter(entry.identifiers))[1:]

```

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,296 @@
+                else:
+                    entry_data[CONF_DEVICES][event_code] = options
+        self.hass.config_entries.async_update_entry(self._config_entry, data=entry_data)
+        self.hass.async_create_task(async_update_options(self.hass, self._config_entry))
```
```suggestion
        self.hass.async_create_task(
            self.hass.config_entries.async_reload(self._config_entry.entry_id)
        )
```

MartinHjelmare commented on homeassistant/components/rfxtrx/config_flow.py:
```diff
@@ -1,12 +1,296 @@
+        self.hass.config_entries.async_update_entry(self._config_entry, data=entry_data)
+        self.hass.async_create_task(async_update_options(self.hass, self._config_entry))
+
+
+async def async_update_options(hass, config_entry: ConfigEntry):
+    """Update options."""
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
```
```suggestion
```

MartinHjelmare commented:
Some final clean up. Then we can merge.

RobBie1221 committed:
Final cleanup

MartinHjelmare reviewed (approved):
Thanks!

Can be merged when build passes. I've restarted CI. Let's see.
